erpnext/erpnext_integrations/taxjar_integration.py

```diff
@@ -364,28 +378,69 @@ def sanitize_error_response(response):
 
 	return response
 
+def get_doc_discount(doc):
+	""""
+		It will take doc discount which is applied on Net Total/ Grand Total. It will distribute discount to line item based on qty
+	"""
+	discount = 0
+	qty = 0
+	for item in doc.items:
+		qty = qty + item.get('qty')
 
-def get_line_items(doc):
-	if not doc.items:
-		return
+	if doc.discount_amount:
+		discount = doc.discount_amount/qty
+	return discount
 
-	discout_percent = 0.0
+def get_line_item_dict(item, distributed_discount, refund_transaction=False):
+	line_item_discount = 0
 
-	if doc.discount_amount:
-		discout_percent = ((doc.discount_amount / doc.total) * 100)
+	# this condition added to prevent item going to negative if distributed discount applied.
+	if item.get("rate") > item.get("price_list_rate"):
+		frappe.throw(_("Row.{0}: Item {1} Rate {2} should not be more than Price List Rate {3}".format(item.get('idx'), item.get('item_name'), item.get('rate'), item.get('price_list_rate'))))
 
-	line_items = []
-	for item in doc.items:
-		if discout_percent:
-			item.discount_amount = ((item.price_list_rate * discout_percent) / 100)
-		line_item = {
-			'id': item.name,
-			'description': item.item_name,
-			'quantity': item.qty,
-			'unit_price': item.price_list_rate,
-			'discount': round(item.discount_amount * item.qty)
-		}
-		line_items.append(line_item)
-
-	return line_items
+	if item.get("discount_percentage"):
+		line_item_discount = (item.get("qty") * (item.get("price_list_rate") * item.get("discount_percentage"))/100)
+	else:
+		line_item_discount = item.get("discount_amount")
+
+	return dict(
+		id = item.get('idx'),
+		description = item.get("item_name"),
+		quantity = abs(item.get("qty")) if refund_transaction else item.get("qty"),
+		unit_price = -abs(item.get("price_list_rate")) if refund_transaction else item.get("price_list_rate"),
+		product_tax_code = item.get("product_tax_code"),
+		sales_tax = item.get("tax_collectable"),
+		discount = line_item_discount + distributed_discount
+	)
+
+def get_line_item_dict_proportionally(item, doc, refund_transaction=False):
+	line_item_discount = 0
+	distributed_discount_percentage = 0
+	distributed_discount = 0
+
+	# this condition added to prevent item going to negative if distributed discount applied.
+	if item.get("rate") > item.get("price_list_rate"):
+		frappe.throw(_("Row.{0}: Item {1} Rate {2} should not be more than Price List Rate {3}".format(item.get('idx'), item.get('item_name'), item.get('rate'), item.get('price_list_rate'))))
+
+	# line item discounting calculate
+	if item.get("discount_percentage"):
+		line_item_discount = (item.get("qty") * (item.get("price_list_rate") * item.get("discount_percentage"))/100)
+	else:
+		line_item_discount = item.get("discount_amount")
 
+	# if paraent level discount aviliable then distribute to line item proportionally
+	if doc.discount_amount:
+		distributed_discount_percentage = (doc.discount_amount / (doc.discount_amount + doc.net_total)) * 100
```

Comment:
can you explain how this works?
Whenever we have any math involved it really helps to write up why we are using any formula and how it works to be nice to the next dev looking at the codebase. And me because I don't understand the formula 🙃 

I was thinking you would grab the scalar of the item's total in regards to the order total. Something like(untested just from my head):
```py3
# total up item and include its line discount so we don't go negative
item_total = (rate * qty) - item_discount

# calc scalar value from order_total minus item_total so we know its contribution to the overall total
item_scalar = (order_total - item_total) / item_total

# now that we know the scale of the total we can take a piece of the order discount
item_distributed_discount = order_discount * item_scalar
```

Reply:
Which Part did you not understand @DeveloperMetal 
Lets take e.g . Test Item with 100 price list rate and qty is 1 = 100
line item discount = 20% = rate will be 80  so amount would be = 80

lets we have shipping 100 and sales tax  7.5 so grand total = 187.5
and we provide a discount 20% on grand total then sales tax reduce to 3.98 because it calculate itself so parent level discount would be = 183.98 *20% = 36.80

so this amount contains shipping discount and line item discount so line 433 convert to distributed discount percentage which is not 20% but 36.80 /(36.80 + 43.20) *100 = 46%

so we distribute 46% to line item so line no 436 which contain rate = 80 * 46% = 36.80
distributed discount = 20+36.80 = 56.8
so your sub_order total = net total + shipping = 43.20 + 100 = 143.20  and 
taxjar validation should be line item total + shipping - discount should match with sub order amount

taxjar : 
A :sub order amount = net_total + shipping = 43.20 + 100 = 143.20

B: line item total +shipping -discount = 100+100-56.8 = 143.20 

taxjar validation paramater A should equal to B= 143.20 == 143.20

Reply:
if I take your suggested code
# total up item and include its line discount so we don't go negative
item_total = (price_list_rate * qty) - item_discount
item_total = (100 * 1) - 20 = 80
# calc scalar value from order_total minus item_total so we know its contribution to the overall total
item_scalar = (order_total - item_total) / item_total
item_scalar = (180 - 80) / 80 = 1.25
# now that we know the scale of the total we can take a piece of the order discount
item_distributed_discount = order_discount * item_scalar 
item_distributed_discount = 36.80 * 1.25 = 46%
distributed discount = 20+36.80 = 56.8
so your sub_order total = net total + shipping = 43.20 + 100 = 143.20 and
taxjar validation should be line item total + shipping - discount should match with sub order amount

taxjar :
A :sub order amount = net_total + shipping = 43.20 + 100 = 143.20

B: line item total +shipping -discount = 100+100-56.8 = 143.20

taxjar validation paramater A should equal to B= 143.20 == 143.20

Reply:
@vishdha I meant specifically line 433, this part `(doc.discount_amount / (doc.discount_amount + doc.net_total)) * 100`

How did you arrive to a one line calculation? what made you take the discount_amount and divide by discount + net_total. Don't need value examples just the how did you get that equation.

On my example I was just pointing out at the parts of the equation which is why I broke it down into 3 parts so its easier to read.